comment/src/comment.ts

```diff
@@ -0,0 +1,280 @@
+/// The block-comment tokens are defined on a language basis.
+export const toggleBlockCommentCmd: Command = view => {
+  return dispatchToggleComment(toggleBlockComment(CommentOption.Toggle), view)
+}
+
+/// Comments the current `SelectionRange` using block-comments.
+/// The block-comment tokens are defined on a language basis.
+export const blockCommentCmd: Command = view => {
+  return dispatchToggleComment(toggleBlockComment(CommentOption.OnlyComment), view)
+}
+
+/// Uncomments the current `SelectionRange` using block-comments.
+/// The block-comment tokens are defined on a language basis.
+export const blockUncommentCmd: Command = view => {
+  return dispatchToggleComment(toggleBlockComment(CommentOption.OnlyUncomment), view)
+}
+
+const dispatchToggleComment = (cmd: (st: EditorState) => Transaction | null, view: EditorView): boolean => {
+  let tr = cmd(view.state)
+  if (!tr) return false
+  view.dispatch(tr)
+  return true
+}
+
+export enum CommentOption {
+  Toggle,
+  OnlyComment,
+  OnlyUncomment,
+}
+
+export const toggleBlockComment = (option: CommentOption) => (state: EditorState): Transaction | null => {
+  type BlockCommentData = { blockComment: { open: string, close: string } | undefined } | undefined
+  const data = state.languageDataAt<BlockCommentData>("commentTokens", state.selection.primary.from)[0]
+  return data === undefined || data.blockComment === undefined || data.blockComment.open === undefined || data.blockComment.close === undefined
+    ? null
+    : new BlockCommenter(data.blockComment.open, data.blockComment.close).toggle(option, state)
+}
+
+export const toggleLineComment = (option: CommentOption) => (state: EditorState): Transaction | null => {
+  type LineCommentData = { lineComment: string | undefined } | undefined
+  const data = state.languageDataAt<LineCommentData>("commentTokens", state.selection.primary.from)[0]
+  return data === undefined || data.lineComment === undefined
+    ? null
+    : new LineCommenter(data.lineComment).toggle(option, state)
+}
+
+// This class performs toggle, comment and uncomment
+// of block comments in languages that support them.
+// The `open` and `close` arguments refer to the open and close
+// tokens of which this `BlockCommenter` is made up.
+export class BlockCommenter {
+  open: string
+  close: string
+  margin: string
+  constructor(open: string, close: string, margin: string = " ") {
+    this.open = open
+    this.close = close
+    this.margin = margin
+  }
+
+  toggle(option: CommentOption, state: EditorState): Transaction | null {
+    const selectionCommented = this.isSelectionCommented(state)
+    if (selectionCommented !== null) {
+      if (option !== CommentOption.OnlyComment) {
+        const tr = state.t()
+        const mapRef = tr.mapRef()
+        for (const range of selectionCommented) {
+          const open = range.open
+          const close = range.close
+          open.pos = mapRef.mapPos(open.pos)
```

Comment:
Did you try the array-of-changes overload on `tr.change` instead of manual mapping?

Reply:
Yes, I tried. The main issue with the `tr.change` overload is about cursor positioning. For block comments, it's important to keep cursor "inside" the comments (behaviour of VSCode). But more importantly, having the cursor outside the comments feels odd to the user.